eth/catalyst/simulated_beacon.go

```diff
@@ -0,0 +1,247 @@
+// Copyright 2023 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package catalyst
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/beacon/engine"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+// withdrawalQueue implements a FIFO queue which holds withdrawals that are pending inclusion
+type withdrawalQueue struct {
+	// pending holds withdrawals that will be included in a future block
+	pending []*types.Withdrawal
+	mu      sync.Mutex
+}
+
+// queued returns withdrawals which are pending inclusion in the next block
+func (w *withdrawalQueue) queued() []*types.Withdrawal {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	var queueCount int
+	if len(w.pending) >= 10 {
+		queueCount = 10
+	} else {
+		queueCount = len(w.pending)
+	}
+
+	p := make([]*types.Withdrawal, queueCount)
```

Comment:
Can replace the above with:
```suggestion
        count := min(10, len(w.pending))
	p := make([]*types.Withdrawal, count)
```

Reply:
`min` isn't a function though?

Reply:
I think it would be nice. They're being added to go 1.21 so we can always remove it after a few more releases.

Reply:
This is no longer relevant after the refactor I pushed today.